### PR TITLE
Thread safety for URLSession

### PR DESCRIPTION
**Pull request by johnno1962**
Hi Apple,

I’ve been seeing assertion failures using a URLSession from multiple threads on [this line](https://github.com/apple/swift-corelibs-foundation/blob/master/Foundation/URLSession/TaskRegistry.swift#L63). This PR makes a small change so allocation of the taskIdentifier in URLSession is thread safe.

Thanks.

### CODE REVIEW

**Comment by alblue** [comment]
@swift-ci please test

File: Foundation/URLSession/URLSession.swift [hunk 1]
```diff
@@ -192,6 +192,7 @@ public let NSURLSessionTransferSizeUnknown: Int64 = -1
 open class URLSession : NSObject {
     fileprivate let _configuration: _Configuration
     fileprivate let multiHandle: _MultiHandle
+    fileprivate let taskIdentifierLock = NSLock()
```
Comment (weissi) [inline Foundation/URLSession/URLSession.swift]: my only question is what is the preferred way, `NSLock` or a dispatch queue? I think we should be consistent. @phausler / @parkera do we have a preferred option?

Comment (weissi) [inline Foundation/URLSession/URLSession.swift]: also: the `sessionDescription` and `delegate` properties are also mutable and accessible from the outside. I suppose the should be locked too.

Comment (johnno1962) [reply Foundation/URLSession/URLSession.swift]: I’ve switched to using the workQueue for consistency

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by pushkarnk** [comment]
@swift-ci please test and merge

**Comment by pushkarnk** [comment]
cc @saiHemak

**Comment by johnno1962** [comment]
Thanks @pushkarnk, Are we any closer to merging https://github.com/apple/swift-corelibs-foundation/pull/1198? Is there anything more I can do?

**Pull request merged**